keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md

```diff
@@ -0,0 +1,902 @@
+- [x] (R) Production readiness review completed
+- [x] (R) Production readiness review approved
+- [x] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With Dynamic Resource Allocation (DRA), DRA drivers publish information about
+the devices that they manage in ResourceSlices. This information is used by the
+scheduler when selecting devices for user requests in ResourceClaims.
+
+With this KEP, DRA drivers can define metadata in mixins separately from specific
+devices and include them in a device by reference. This reduces the duplication
+in ResourceSlices and allows for more compact device definitions. It also enables
+defining devices with more attributes, capacities and consumed counters than without
+mixins. Mixins can also be used in counter sets.
+
+## Motivation
+
+DRA requires that drivers publish all available devices on a node/cluster in
+`ResourceSlice` objects. There are scenarios where the number of devices
+can be pretty large and each device might have a relatively large amount
+of metadata associated with it, primarily in the form of attributes and
+capacity. This has a few consequences:
+
+* Several of the devices might have similar metadata, resulting in a lot of
+  duplication between the published devices.
+* The size of the data required to specify each device reduces the number of
+  devices that can be defined in a single ResourceSlice without hitting the
+  Kubernetes size limit.
+
+The latter can be addressed by splitting the devices across multiple
+`ResourceSlice`s within a single pool, but that isn't always an option.
+In particular, DRA currently doesn't allow sharing counters across `ResourceSlice`s,
+meaning that the number of devices that can fit into a single `ResourceSlice`
+also limits the number of partitionable devices for a single physical device.
+
+### Goals
+
+- Enable a more compact way to define devices in ResourceSlices so duplication can
+  be reduced and a larger number of devices can be published within a single
+  ResourceSlice.
+- Enable defining devices with more attributes, capacities, and consumed counters.
+- Enable defining counter sets with more counters.
```

Comment:
Two things we should consider (not sure if these are goals or implementation choices): 1) enabling mixins to be per-pool not per-resource slice; 2) enabling counters to be per-pool not per-resource slice.

If necessary, these could be considered in beta. But I do think we're going to need them in time. The second may belong in partitionable not here.

Reply:
Yeah, we have an issue for that in https://github.com/kubernetes/kubernetes/issues/130785. We definitely need to make a decision on this in this cycle, as I think changing this must happen over two releases. I was hoping to handle this separately from this KEP, but open to including it here.